core/src/main/scala/cats/syntax/semigroup.scala

```diff
@@ -1,14 +1,16 @@
 package cats
 package syntax
 
+import cats.macros.Ops
+
 trait SemigroupSyntax {
   // TODO: use simulacrum instances eventually
   implicit def semigroupSyntax[A: Semigroup](a: A): SemigroupOps[A] =
     new SemigroupOps[A](a)
 }
 
-class SemigroupOps[A](lhs: A)(implicit A: Semigroup[A]) {
-  def |+|(rhs: A): A = A.combine(lhs, rhs)
-  def combine(rhs: A): A = A.combine(lhs, rhs)
-  def combineN(rhs: Int): A = A.combineN(lhs, rhs)
+class SemigroupOps[A: Semigroup](lhs: A) {
+  def |+|(rhs: A): A = macro Ops.binop[A, A]
```

Comment:
Consider the following generic code using semigroup:

```
def sum[A: Semigroup](as: Iteratable[A]): A =
  as.foldLeft(0)(_ |+| _)
```

Say I pass it a `Vector[Int]` with 100,000 items. With the old code, I would instantiate 100k instances of `SemigroupOps`, one for each step in the `foldLeft`. With the new code, I would allocate 0 instances of `SemigroupOps`; the code would be equivalent to:

```
def sum[A](as: Iteratable[A])(implicit ev: Semigroup[A]): A =
  as.foldLeft(0)(ev.combine)
```

So it is an efficiency improvement that removes the penalty of using Ops implicits rather than explicitly threading and using the type class instances.